### PR TITLE
C++: Speed up PrintAST.

### PR DESCRIPTION
**Pull request by criemen**
Performance numbers:
Using 16 threads for evaluator, with empty cache
boost, lockpool.cpp
before: total 85s, qlClass 10s, EB 30s
after: total 75s, qlClass 3s

openjdk, ad_x86.cpp
before: total 93s, qlClass 11s, EB 28
after: total 81s, qlClass 7.9s

The times for the individual predicates are from the query log, the total time is the wall time from the parallel execution.
EB is the predicate in `ElementBase` that computes the `el` that need the default implementation of `getAPrimaryQlClass`, i.e. where the predicate is not overriden.
Note: Changing this default from "???" to no result is a breaking API change.

Part of https://github.com/github/codeql-c-analysis-team/issues/129

### CODE REVIEW

**Comment by criemen** [comment]
New performance numbers, see first post.
Unfortunately the first version pruned a bit too much, and was even faster, but incorrect.

File: cpp/ql/src/semmle/code/cpp/Element.qll [hunk 1]
```diff
@@ -65,11 +65,10 @@ class ElementBase extends @element {
    * which they belong; for example, `AddExpr` is a primary class, but
    * `BinaryOperation` is not.
    *
-   * This predicate always has a result. If no primary class can be
-   * determined, the result is `"???"`. If multiple primary classes match,
-   * this predicate can have multiple results.
+   * This predicate can have multiple results if multiple primary classes match.
+   * For some elements, this predicate may not exist.
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/Element.qll:69]:
```suggestion
   * For some elements, this predicate may not have a result.
```

**Review by jbj — COMMENTED** [review]
Something's a bit dodgy with the performance numbers you posted:

>  boost, lockpool.cpp
> before: total 85s, qlClass 10s, EB 30s
> after: total 75s, qlClass 3s

37 seconds are removed from the predicates you're tracking, but only 10 seconds are removed from the total. So there must be other predicates that are worth tracking.

**Comment by criemen** [comment]
The performance numbers are quite weird tbh - maybe I am using too high parallelism to really see the effects of the optimization here.
We can discuss this in our call tomorrow.

**Comment by criemen** [comment]
I have now new performance numbers taking with a single-threaded execution (and cleared cache) on boost:
boost 1thread before 130s, eb 22.7s, qlclass 8.7s

boost 1thread after 88s, eb 0s, qlclass 1.5s

Showing a considerable speedup in the two predicates that are modified in this PR.

**Review by jbj — APPROVED** [review]

**Pull request merged**